### PR TITLE
Allow expansion service to choose pickler.

### PR DESCRIPTION
**Pull request by robertwb**
This needs to be added here as it is a global option, not a pipeline option.
This is useful as dill has some issues with objects created from exec/eval
which are used for remotely-defined PythonCallable objects.

Also refactored the pickler choice logic a bit for clarity and better
error checking.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @chamikaramj It'd be good to get this in before the release cut--possibly considering changing the default (for Python-from-Java).

File: sdks/python/apache_beam/internal/pickler.py [hunk 1]
```diff
@@ -68,10 +68,12 @@ def load_session(file_path):
 def set_library(selected_library=DEFAULT_PICKLE_LIB):
   """ Sets pickle library that will be used. """
   global desired_pickle_lib
-  if selected_library == USE_DILL and desired_pickle_lib != dill_pickler:
+  if (selected_library == USE_DILL) != (desired_pickle_lib == dill_pickler):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/internal/pickler.py:71]: Did you mean "(selected_library == USE_DILL) **and**"

Logic is bit hard to follow.

**Comment by chamikaramj** [comment]
LGTM to merge after verifying the logic. Thanks.

**Comment by robertwb** [comment]
Run Python_PVR_Flink PreCommit

File: sdks/python/apache_beam/internal/pickler.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/internal/pickler.py:72]: If the new value of (pickler is dill) is different than the old value, we have to set (or unset) the hooks. Added a comment to clarify.

**Comment by tvalentyn** [comment]
cc: @ryanthompson591

File: sdks/python/apache_beam/runners/portability/local_job_service_main.py [hunk 2]
```diff
@@ -129,6 +131,9 @@ def run(argv):
     print('Server started at port', port)
     return
 
+  if options.default_pickler:
+    pickler.set_library(options.default_pickler)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: should we set the `--pickler_library` pipeline option in this case?

Also noting the naming difference for a similar concept - should this be renamed into `--pickler_library`?

Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: I wasn't aware that there was such a pipeline option (and it doesn't seem possible to set on a per-pipeline basis). 

If you feel it's worth changing the name, we should do so before the next release, though we'd have to make sure it gets set in both places (setting in stager seems too late).

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: > I wasn't aware that there was such a pipeline option
It's used for example here: https://github.com/apache/beam/blob/f0cd27596e3a59d6ee013ee2db232d24f46ad70a/sdks/python/apache_beam/runners/worker/sdk_worker_main.py#L103

> it doesn't seem possible to set on a per-pipeline basis
I wonder if this is the same as: https://github.com/apache/beam/issues/21615

> If you feel it's worth changing the name, we should do so before the next release
I think it would be worthwhile. Would someone working on xlang have bandwidth look into that?

If we are making cloudpickle default, I would consider tightening the bounds on cloudpickle to 
cloudpickle>=2.2.0,<2.3.0 , and bumping that bound every beam release until we decide to vendor the library.

Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: I can take on this change. Should we just re-use `--pickle_library` like the pipeline option? (Actually fixing #21615 is out of scope for now.) It'd be another cherry-pick into the release branch.

Tightening the bounds feels like it should be a separate PR.

Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: https://github.com/apache/beam/pull/23111

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: >  It'd be another cherry-pick into the release branch.

Cherrypick, assuming #23111 is merged as is: https://github.com/apache/beam/pull/23118

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/portability/local_job_service_main.py:135]: > Tightening the bounds feels like it should be a separate PR.

Sending https://github.com/apache/beam/pull/23120
